NWNXLib/Log.cpp

```diff
@@ -124,21 +124,46 @@ void WriteToLogFile(const char* message)
     }
 }
 
-// Unfortunately this needs to be backed by an std::string rather than a const char* because the C++ standard library
-// doesn't provide a hashing function for a const char* and I'm too lazy to write or copy/paste one.
-// Small perf hit when logging.
-static std::unordered_map<std::string, Channel::Enum> s_LogLevelMap;
+template <typename T>
+struct CharpEqualTo : public std::binary_function<T, T, bool>
+{
+    bool operator()(const T& x, const T& y) const
+    {
+        return !strcmp(x, y);
+    }
+};
+
+struct CharpHash
+{
+    int operator()(const char * s) const
+    {
+        int i = 0, h = 0;
+        while (s[i]) {
+            h += s[i];
+            i++;
+        }
+        return h | (i << 20);
+    }
+};
+
+static std::unordered_map<const char*, Channel::Enum, CharpHash, CharpEqualTo<const char*>> s_LogLevelMap;
+static Channel::Enum s_LogLevelMax;
 
 Channel::Enum GetLogLevel(const char* plugin)
 {
-    // TODO: Is this thread safe? I think so if we're just looking up but I don't know.
     auto entry = s_LogLevelMap.find(plugin);
     return entry == std::end(s_LogLevelMap) ? Channel::SEV_NOTICE : entry->second;
 }
 
+Channel::Enum GetMaxLogLevel()
+{
+    return s_LogLevelMax;
+}
+
 void SetLogLevel(const char* plugin, Channel::Enum logLevel)
 {
     s_LogLevelMap[plugin] = logLevel;
+    s_LogLevelMax = std::max(s_LogLevelMax, logLevel);
```

Comment:
Actually, this should to be:
```
for (auto it : s_LogLevelMap)
    s_LogLevelMax = std::max(s_LogLevelMax, it->second);
```
because levels can change dynamically. Although since that only affects performance if the max drops, we can do it in a follow on.